cpp/src/quantiles/tdigest/tdigest.cu

```diff
@@ -0,0 +1,356 @@
+      return;
+    }
+    right_index++;
+  } else {
+    if (left_index == 0) {
+      output[tid] = cast_op(lerp(*min_val, mean[left_index], diff / (weight[left_index] / 2)));
+      return;
+    }
+    left_index--;
+    diff += weight[left_index] / 2 + weight[right_index] / 2;
+  }
+
+  diff /= (weight[left_index] / 2 + weight[right_index] / 2);
+  output[tid] = cast_op(lerp(mean[left_index], mean[right_index], diff));
+}
+
+/**
+ * @brief Calculate approximate percentiles on a provided tdigest column.
+ *
+ * Produces a LIST column where each row N represents output from querying the
+ * corresponding tdigest of from row N in `input`. The length of each output list
+ * is the number of percentiles specified in `percentiles`
+ *
+ * @param input           tdigest input data. One tdigest per row.
+ * @param percentiles     Desired percentiles in range [0, 1].
+ * @param stream          CUDA stream used for device memory operations and kernel launches
+ * @param mr              Device memory resource used to allocate the returned column's device
+ * memory
+ *
+ * @returns Column of doubles containing requested percentile values.
+ */
+template <typename T>
+std::unique_ptr<column> compute_approx_percentiles(structs_column_view const& input,
+                                                   column_view const& percentiles,
+                                                   data_type output_type,
+                                                   rmm::cuda_stream_view stream,
+                                                   rmm::mr::device_memory_resource* mr)
+{
+  lists_column_view lcv(input.child(tdigest_centroid_column_index));
+  column_view min_col = input.child(tdigest_min_column_index);
+  column_view max_col = input.child(tdigest_max_column_index);
+
+  // offsets, representing the size of each tdigest
+  auto offsets = lcv.offsets();
+
+  // extract means and weights
+  auto data = lcv.parent().child(lists_column_view::child_column_index);
+  structs_column_view tdigest(data);
+  auto mean   = tdigest.child(tdigest_mean_column_index);
+  auto weight = tdigest.child(tdigest_weight_column_index);
+
+  // compute summed weights
+  auto cumulative_weights = cudf::make_fixed_width_column(data_type{type_id::FLOAT64},
+                                                          mean.size(),
+                                                          mask_state::UNALLOCATED,
+                                                          stream,
+                                                          rmm::mr::get_current_device_resource());
+  auto keys               = cudf::detail::make_counting_transform_iterator(
+    0,
+    [offsets = offsets.begin<offset_type>(), num_offsets = offsets.size()] __device__(size_type i) {
+      return static_cast<size_type>(
+               thrust::upper_bound(thrust::seq, offsets, offsets + num_offsets, i) - offsets) -
+             1;
```

Comment:
This is somewhat of a personal preference, but I often see this pattern when using `upper_bound/find` functions and it can be difficult to distinguish the math that is being applied to the result of the `upper_bound` call from the functions passed to `upper_bound` or even notice that math is happening. 

This is how I'd write this to be more explicit and less likely to be overlooked. 
```suggestion
      return thrust::distance(thrust::prev(thrust::upper_bound(thrust::seq, offsets, offsets + num_offsets, i)), offsets_begin);
```
Or break it up into pieces to make it a bit more digestable:
```
      auto const iter = thrust::prev(thrust::upper_bound(thrust::seq, offsets, offsets + num_offsets, i))
      return thrust::distance(iter, offsets_begin);
```